distributed/dashboard/tests/test_scheduler_bokeh.py

```diff
@@ -610,6 +611,41 @@ async def test_TaskGraph_order(c, s, a, b):
     assert gp.node_source.data["state"][gp.layout.index[y.key]] == "erred"
 
 
+####Adding TGroupGraph tests ######
+@gen_cluster(client=True)
+async def test_TGroupGraph(c, s, a, b):
+    tgg = TGroupGraph(s)
+
+    futures = c.map(slowinc, range(10), delay=0.001)
+    await wait(futures)
+
+    tgg.update()
+    dnode = dict(tgg.node_source.data)
+    assert all(len(L) == 1 for L in dnode.values())
+    assert dnode["name"] == ["slowinc"]
+    assert dnode["tot_tasks"] == [10]
+
+    darr = dict(tgg.arrow_source.data)
+    assert all(len(L) == 0 for L in darr.values())
+
+    futures2 = c.map(dec, range(5))
+    await wait(futures2)
+
+    tgg.update()
+    dnode = dict(tgg.node_source.data)
+    assert all(len(L) == 2 for L in dnode.values())
+    assert dnode["name"] == ["dec", "slowinc"]  # they follow the tg_stack
+    assert dnode["tot_tasks"] == [5, 10]
+
+    del futures, futures2
+
+    tgg.update()  ###for some reason after deleting the futures the tgg.node_source.data.values are not clear.
+    assert not all(tgg.node_source.data.values())
```

Comment:
```suggestion
    assert not any(tgg.node_source.data.values())
```

Reply:
Problem solved after James commit, thanks @jrbourbeau for helping with the debugging .